### ISSUE TITLE
INBOUND and OUTBOUND nodes/links for small subsets - Network Export

### ISSUE THREAD

**Issue by msdogan** [ucd-cws/calvin-network-tools#49]
@jrmerz we have a small issue with exporting matrix for a small subset of network. `cnf matrix` command is missing boundary conditions for nodes. You know when we create small subset, it takes solutions from a previous feasible run, stored in HOBBES, and puts as boundary if network is less than 82-year period. For reservoir nodes, it is fine. We have `INITIAL` and `FINAL` nodes that serve as boundary conditions, but for example if I create a 2-node subset, it does not boundary for `D5` in the following example, which creates mass balance issue since all nodes (except for `SOURCE` and `SINK`) must have at least one incoming and outgoing links.

command:
```
AD3+msdogan@CWS-Amargosa  ~/Desktop/GitHub
$ cnf matrix --verbose --format=csv --start=2002-10 --stop=2003-1 --ts=. --to=network --outnodes=nodes --max-ub=1000000000000 SR_SHA D5
```

network matrix:
```
SOURCE	INITIAL	0	0.0	1.0	0.0	1000000000000.0
INITIAL	SR_SHA.2002-10-31	0	0.0	1.0	1909.535	1909.535
SR_SHA.2002-10-31	SR_SHA.2002-11-30	0	-7.003439999999999	0.99744	630.403	630.403
SR_SHA.2002-10-31	SR_SHA.2002-11-30	1	-2.97378	0.99744	737.3889999999999	737.3889999999999
SR_SHA.2002-10-31	SR_SHA.2002-11-30	2	-1.46632	0.99744	332.208	2032.2079999999999
SR_SHA.2002-11-30	SR_SHA.2002-12-31	0	-6.97197	0.99859	609.442	609.442
SR_SHA.2002-11-30	SR_SHA.2002-12-31	1	-3.05553	0.99859	700.8380000000001	700.8380000000001
SR_SHA.2002-11-30	SR_SHA.2002-12-31	2	-1.47896	0.99859	389.72	1941.72
FINAL	SINK	0	0.0	1.0	0.0	1000000000000.0
SR_SHA.2002-12-31	FINAL	0	0.0	1.0	2661.7929999999997	2661.7929999999997
SOURCE	INFLOW.2002-10-31	0	0.0	1.0	0.0	1000000000000.0
INFLOW.2002-10-31	SR_SHA.2002-10-31	0	0.0	1.0	187.669	187.669
SOURCE	INFLOW.2002-11-30	0	0.0	1.0	0.0	1000000000000.0
INFLOW.2002-11-30	SR_SHA.2002-11-30	0	0.0	1.0	229.032	229.032
SOURCE	INFLOW.2002-12-31	0	0.0	1.0	0.0	1000000000000.0
INFLOW.2002-12-31	SR_SHA.2002-12-31	0	0.0	1.0	1093.016	1093.016
SR_SHA.2002-10-31	D5.2002-10-31	0	-0.7673800000000001	1.0	0.0	332.033
SR_SHA.2002-10-31	D5.2002-10-31	1	-0.73292	1.0	0.0	166.017
SR_SHA.2002-10-31	D5.2002-10-31	2	-0.72938	1.0	0.0	166.017
SR_SHA.2002-10-31	D5.2002-10-31	3	-0.7255199999999999	1.0	0.0	166.017
SR_SHA.2002-10-31	D5.2002-10-31	4	-0.72124	1.0	0.0	166.017
SR_SHA.2002-10-31	D5.2002-10-31	5	-0.7155199999999999	1.0	0.0	166.017
SR_SHA.2002-10-31	D5.2002-10-31	6	0.0	1.0	0.0	1000000000000.0
SR_SHA.2002-11-30	D5.2002-11-30	0	-0.77574	1.0	0.0	321.695
SR_SHA.2002-11-30	D5.2002-11-30	1	-0.7404	1.0	0.0	160.84799999999998
SR_SHA.2002-11-30	D5.2002-11-30	2	-0.7378399999999999	1.0	0.0	160.84799999999998
SR_SHA.2002-11-30	D5.2002-11-30	3	-0.73551	1.0	0.0	160.847
SR_SHA.2002-11-30	D5.2002-11-30	4	-0.73237	1.0	0.0	160.84799999999998
SR_SHA.2002-11-30	D5.2002-11-30	5	-0.72648	1.0	0.0	160.84799999999998
SR_SHA.2002-11-30	D5.2002-11-30	6	0.0	1.0	0.0	1000000000000.0
SR_SHA.2002-12-31	D5.2002-12-31	0	-0.77895	1.0	0.0	332.033
SR_SHA.2002-12-31	D5.2002-12-31	1	-0.7746	1.0	0.0	166.017
SR_SHA.2002-12-31	D5.2002-12-31	2	-0.7703399999999999	1.0	0.0	166.017
SR_SHA.2002-12-31	D5.2002-12-31	3	-0.76563	1.0	0.0	166.017
SR_SHA.2002-12-31	D5.2002-12-31	4	-0.7568699999999999	1.0	0.0	166.017
SR_SHA.2002-12-31	D5.2002-12-31	5	-0.74775	1.0	0.0	166.017
SR_SHA.2002-12-31	D5.2002-12-31	6	0.0	1.0	0.0	1000000000000.0
```

A previously exported example with `INBOUND` and `OUTBOUND` nodes and links:
```
SOURCE	INBOUND.2001-08-31	0	0	1	0	1000000
INBOUND.2001-08-31	SR_CLE.2001-08-31	0	0	1	230.580612	230.580612
OUTBOUND.2001-08-31	SINK	0	0	1	0	1000000
D94.2001-08-31	OUTBOUND.2001-08-31	0	0	1	230.580612	230.580612
SOURCE	INBOUND.2001-09-30	0	0	1	0	1000000
INBOUND.2001-09-30	SR_CLE.2001-09-30	0	0	1	213.177673	213.177673
OUTBOUND.2001-09-30	SINK	0	0	1	0	1000000
D94.2001-09-30	OUTBOUND.2001-09-30	0	0	1	213.177673	213.177673
SR_CLE.2001-08-31	D94.2001-08-31	0	-0.8009061426161982	1	0	71.940552
SR_CLE.2001-08-31	D94.2001-08-31	1	-0.7933839317774485	1	0	35.970276
SR_CLE.2001-08-31	D94.2001-08-31	2	-0.7904757806139715	1	0	35.970276
SR_CLE.2001-08-31	D94.2001-08-31	3	-0.7863525091203776	1	0	35.970275000000015
SR_CLE.2001-08-31	D94.2001-08-31	4	-0.6791794424930186	1	0	35.970275999999984
SR_CLE.2001-08-31	D94.2001-08-31	5	-0.6013748129149743	1	0	35.97027600000001
SR_CLE.2001-08-31	D94.2001-08-31	6	0	1	0	1000000
SR_CLE.2001-09-30	D94.2001-09-30	0	-0.7420960355688229	1	0	69.700592
SR_CLE.2001-09-30	D94.2001-09-30	1	-0.7350739253575176	1	0	34.850288000000006
SR_CLE.2001-09-30	D94.2001-09-30	2	-0.7327835074265064	1	0	34.850303999999994
SR_CLE.2001-09-30	D94.2001-09-30	3	-0.7305389589254676	1	0	34.850280999999995
SR_CLE.2001-09-30	D94.2001-09-30	4	-0.7269706403641447	1	0	34.850296000000014
SR_CLE.2001-09-30	D94.2001-09-30	5	-0.7197610889732475	1	0	34.850295999999986
SR_CLE.2001-09-30	D94.2001-09-30	6	0	1	0	1000000
```

**Comment by jrmerz** [comment]
@msdogan think I found the issue.  Please pull and give it a test.

**Comment by msdogan** [comment]
@jrmerz the command still does not put `INBOUND` and `OUTBOUND` nodes/links

**Comment by jrmerz** [comment]
did you run from source (node ./bin/cnf.bin.js ... )?  I have not pushed update to NPM yet.

**Comment by msdogan** [comment]
nope, I run globally with npm `cnf` let me try again, then

**Comment by msdogan** [comment]
@jrmerz now it worked! Thanks!

**Comment by jrmerz** [comment]
@msdogan great! :)

Pushed to NPM calvin-network-tools@2.0.8

**Comment by msdogan** [comment]
@jdherman @mfefer
now you can run small subset and one-year example, with an infeasible solution, though ( :) sorry), if you are running without debug mode. I believe problem is due to flow differences between CALVIN (S09I05) and PyVIN. HOBBES takes reservoir storage boundary conditions (INITIAL and FINAL) from S09I05 82-year run not from PyVIN, which are slightly different. But it is a data issue not programming.

you can update your calvin-network-tools repo with:

```git pull```
```
AD3+msdogan@CWS-Amargosa  /z/calvin-network-tools (master)
$ git pull
remote: Counting objects: 3, done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 3 (delta 0), reused 0 (delta 0), pack-reused 0
Unpacking objects: 100% (3/3), done.
From https://github.com/ucd-cws/calvin-network-tools
   856ceb4..2f36bc2  master     -> origin/master
Updating 88d5bd0..2f36bc2
Fast-forward
 package.json | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
```

```npm install -g calvin-network-tools```
```
AD3+msdogan@CWS-Amargosa  /z/calvin-network-tools (master)
$ npm install -g calvin-network-tools
npm WARN deprecated node-uuid@1.4.7: use uuid module instead
Z:\nodejs\cnf -> Z:\nodejs\node_modules\calvin-network-tools\bin\cnf.bin.js
Z:\nodejs
`-- calvin-network-tools@2.0.8
  +-- csv-parse@1.2.0
  +-- exit-on-epipe@1.0.0
  +-- hobbes-network-format@2.0.7
  +-- printj@1.0.0
  +-- superagent@1.8.5
  | `-- form-data@1.0.0-rc3
  |   `-- mime-types@2.1.14
  |     `-- mime-db@1.26.0
  `-- xlsx@0.8.2
```

**Comment by msdogan** [comment]
@jrmerz is there a way to store PyVIN output results in a separate HOBBES calvin-network-data repo? And then we can define data path while exporting the matrix.
@jdherman what do you think?

**Comment by jrmerz** [comment]
@msdogan isn't the solution here to apply the results from PyVIN back to the node/link geojson files?  We have this functionality for HEC-PRM in the calvin-network-tools.  But nothing implemented yet for PyVIN results.

**Comment by msdogan** [comment]
@jrmerz yes, but in that case, run results using HEC-PRM as a solver could be infeasible :)

**Comment by jrmerz** [comment]
well I would recommend branching the data.  perhaps moving HEC-PRM off to it's own branch and leaving PyVIN on master.  You can always specify which repo you want to use  in the calvin-network-tools via the --data flag

**Comment by msdogan** [comment]
that sounds like a good option to me

**Comment by jdherman** [comment]
@msdogan thanks. Whatever you guys decide on the data-syncing is fine -- really I'd be fine with leaving it separately in CSV (timeseries) form, I think it's easier to analyze that way. 

The main reason I can see to merge the PyVIN results back into the HOBBES format is if we want to visualize them in the web interface.
